### PR TITLE
[transform] add a disc.cache_read transform primitive

### PR DESCRIPTION
**Pull request by wyzero**
to #787

### CODE REVIEW

File: tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc [hunk 1]
```diff
@@ -575,6 +577,309 @@ DiagnosedSilenceableFailure FoldProducerExtractSliceOp::applyToOne(
...
+        sliceOp.getOperation(),
+        "expect the strides of target slice op are all ones\n");
+
+  // verify the target op have static shape
+  auto targetTy = targetOp->getResult(0).getType().cast<ShapedType>();
+  if (!targetTy.hasStaticShape() || targetTy.getRank() == 0)
+    return mlir::emitDefiniteFailure(
+        targetOp, "expect the targetOp has static shape with rank > 0\n");
+
+  ArrayRef<Operation*> anchorOps = state.getPayloadOps(getAnchor());
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc:819]: newbie: what is an anchor?

File: tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc [hunk 2]
```diff
@@ -575,6 +577,309 @@ DiagnosedSilenceableFailure FoldProducerExtractSliceOp::applyToOne(
...
+
+  if (!permutationVec.empty()) {
+    newOffsets =
+        disc_linalg_ext::interchange<OpFoldResult>(newOffsets, permutationVec);
+    newSizes =
+        disc_linalg_ext::interchange<OpFoldResult>(newSizes, permutationVec);
+  }
+
+  auto newShape = info.dimSizes;
+  newShape = disc_linalg_ext::interchange<int64_t>(newShape, info.permutation);
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc:722]: why only use inner most tiled-dims to do interchange to make new shape?

**Review by qiuxiafei — APPROVED** [review]

File: tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc:819]: anchor op is used to decide where to place the new created `multi_level_pack` op. It's one of the inputs of cache_read op.

File: tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc (hunk 2 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/tools/disc-transform/TransformOps/TransformOpsExt.cc:722]: here we have already sliced a tile from the packed weight, thus only needs to transpose the tile (inner most dims) to replace the result of the original tile slice op.

**Pull request merged**